[Wilfred Wong] iP

Wilfredwongkc opened the pull request:
# DukePro #
> “Your mind is for having ideas, not holding them.” – David Allen ([Source](https://dansilvestre.com/productivity-quotes/))

DukePro frees your mind of having to remember things you need to do. It's,
* text-based
* easy to learn
*  ~~FAST~~ SUPER FAST to use

All you need to do is,
1. download it from [here](https://github.com/Wilfredwongkc/ip/releases/tag/A-Jar).
2. double-click it.
3. add your task.
4. let it manage your task for you :kissing_heart:

And it is **FREE**!

Features:
- [x] Managing tasks
- [x] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

## ##

If you Java programmer, you can use it to practice Java too. Here's the main method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Wilfredwongkc committed:
Added Greeting, echo and exit feature

Wilfredwongkc committed:
Added add and list feature. Encapsulated task.

Wilfredwongkc committed:
Added marking as done feature. Changed naming convention to be the same as the "partial solution given".

Wilfredwongkc committed:
Added ToDos, Events and Deadlines features. Made some other minor changes (e.g. creating function to draw line).

Wilfredwongkc committed:
Moved the initialisation of scanner object outside the while-loop so that it doesn't get reinitialised everytime a new input is passed in.

Wilfredwongkc committed:
Modified "input.txt" by adding some input test cases. Modified "EXPECTED.txt" by adding expected result for the the test case inputs.

Wilfredwongkc committed:
Revert "Modified "input.txt" by adding some input test cases. Modified "EXPECTED.txt" by adding expected result for the the test case inputs."

This reverts commit 72141596078e987fe2ba96f1b5c3308c7f1e9f2c.

Wilfredwongkc committed:
Revert "Revert "Modified "input.txt" by adding some input test cases. Modified "EXPECTED.txt" by adding expected result for the the test case inputs.""

This reverts commit eab7050491c502354d86262dcde4b6a5ea698901.

Wilfredwongkc committed:
Added new line to the bottom of the text files.

Wilfredwongkc committed:
Added exception handling feature for making sure users type a description for todo and making sure that user do not type any recognized commands.

Wilfredwongkc committed:
Added delete feature toghether with additional exception handling for the delete feature. Formatted the replies with tabs to make it look nicer.

Wilfredwongkc committed:
Updated the Input.txt and EXPECTED.txt because I modified the layout of the replies and also to test the exception handling by inputing some wrong inputs.

Wilfredwongkc committed:
Add load and save from file feature

Wilfredwongkc committed:
Add date and times feature

Wilfredwongkc committed:
Merge branch 'branch-Level-7'

Wilfredwongkc committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Deadline.java
#	src/main/java/Event.java

Wilfredwongkc committed:
Fix for additonal problems caused by merge

Wilfredwongkc committed:
Abstract storage class

Wilfredwongkc committed:
Remove uncessary imports and print statements to show file or directory creation

Wilfredwongkc committed:
Abstract task list class

Wilfredwongkc committed:
Abstract ui class

Wilfredwongkc committed:
Abstract parser class

Wilfredwongkc committed:
Add data file for loading and saving

Wilfredwongkc committed:
Organize into packages

Wilfredwongkc committed:
Move files to another directory for package creation

Wilfredwongkc committed:
Add JUnit tests

Wilfredwongkc committed:
Add META-INF file to track

Wilfredwongkc committed:
Add find feature

Wilfredwongkc committed:
Add Javadoc

Wilfredwongkc committed:
Follow coding standard

Wilfredwongkc committed:
Merge branch 'branch-Level-9'

Wilfredwongkc committed:
Merge branch 'branch-A-JavaDoc'

Wilfredwongkc committed:
Merge branch 'branch-A-CodingStandard'

tanhuakun commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,83 @@
+ * It will save the final list into the file when programmes end.
+ */
+public class Duke {
+    static void drawLine() {
+        System.out.println("___________________________________________");
+    }
```
Is this function redundant now that UI can drawLine() too?

tanhuakun commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,15 @@
+        this.input = input;
+    }
+
+    public void parse() {
+
+    }
```
What is the purpose of this method? Perhaps methods with no definition should be in abstract classes.

tanhuakun commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,64 @@
+ * Stores date if the task is a deadline or event.
+ */
+public class Task {
+    protected String description;
+    protected boolean isDone;
+    protected String taskType;
```
Do you need to store a different instance of the same String in every object for task type? It could be better to store it as a static field and override getTaskType() in each subclass.

tanhuakun commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,83 @@
+        System.out.println("___________________________________________");
+    }
+
+    public static void main(String[] args) throws IOException {
```
Should an IOException be allowed to crash the whole program? It could be better to handle it in the functions that could cause the exception.

tanhuakun reviewed (commented):
LGTM. There are two redundant methods but otherwise the code and logic works well enough.

Wilfredwongkc committed:
Merge remote-tracking branch 'origin/add-gradle-support'

ganhongyao commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,29 @@
+    public Deadline(String description, String by) {
+        super(description, "Deadline", by);
+        if (by.contains("/")) {
+            date = LocalDate.parse(by.split("/")[2] + "-" + by.split("/")[1] + "-"
+                    + (Integer.parseInt(by.split("/")[0]) < 10 ? "0" + by.split("/")[0] :
+                    by.split("/")[0]));
```
Would it be better to parse the `LocalDate` the overloaded method that allows you to use a self-defined `DateTimeFormatter`?

You can find the method documentation [here](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/time/LocalDate.html#parse(java.lang.CharSequence,java.time.format.DateTimeFormatter)).

ganhongyao commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,83 @@
+    }
+
+    public static void main(String[] args) throws IOException {
+        String s;
```
Should the variable name be something more descriptive, especially since the variable is referenced rather extensively? It may not be clear what `s` represents.

ganhongyao commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,83 @@
+
+    public static void main(String[] args) throws IOException {
+        String s;
+        int taskNumber;
```
Should the task number be declared and initialised in the respective blocks to keep the scope as small as possible? It seems like it's not being used outside of the `delete` and `done` blocks.

ganhongyao commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,119 @@
+import java.nio.file.Paths;
+import java.nio.file.StandardOpenOption;
+
+import java.util.ArrayList;
+
+import java.util.List;
```
Should the line break between these 2 import statements be removed? Especially since they are from the same package.

ganhongyao commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,119 @@
+     * @throw IOException.
+     */
+    public List<Task> load(Path p) throws IOException {
+        String currContent;
+        String[] parts;
```
Would it be better to declare and initialise the variables within the `for` loop?

You can refer to the coding standard [here](https://se-education.org/guides/conventions/java/intermediate.html#variables).

ganhongyao commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,119 @@
+        String[] parts;
+        ArrayList<Task> taskList = new ArrayList<>();
+        List<String> fileContents = Files.readAllLines(p);
+        for (int i = 0; i< fileContents.size(); i++) {
```
Should there be a whitespace before the operator?
```suggestion
        for (int i = 0; i < fileContents.size(); i++) {
```

ganhongyao commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,119 @@
+                }
+                taskList.add(todo);
+            } else if (parts[0].equals("D")) {
+                Deadline deadline = new Deadline(parts[2],parts[3]);
```
```suggestion
                Deadline deadline = new Deadline(parts[2], parts[3]);
```

ganhongyao commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,119 @@
+                }
+                taskList.add(deadline);
+            } else if (parts[0].equals("E")) {
+                Event event = new Event(parts[2],parts[3]);
```
```suggestion
                Event event = new Event(parts[2], parts[3]);
```

ganhongyao commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,119 @@
+     * @throw IOException.
+     */
+    public void write(List<Task> tasks, Path path) throws IOException {
+        String stringToInsert;
+        Task t;
```
Should these variables be declared at the start of the loop iteration to keep the scope small?

ganhongyao commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,64 @@
+        this.isDone = true;
+    }
+
+    public Boolean getIsDone() {
```
Would it be better to use the primitive type instead? I may have missed out on the reason if this was intended (sorry in advance!)

ganhongyao commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,63 @@
+ * It allows adding and deleting of tasks from the list.
+ */
+public class TaskList {
+    private List<Task> taskList;
```
Would it be better to name this `tasks` instead? It may be confusing for the class name and the field name to be the same.
```suggestion
    private List<Task> tasks;
```

ganhongyao commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,63 @@
+    }
+
+    /**
+     * Add tasks to the list.
```
Should it be "Adds" instead? Same for the other javadoc comments :)
```suggestion
     * Adds tasks to the list.
```

ganhongyao commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,116 @@
+public class Ui {
+    String logo;
+    Storage storage;
+    TaskList list;
```
Should these variables be declared as `private` instead?

ganhongyao reviewed (commented):
I added a few comments with regards to coding standards. Feel free to ping me if there were parts that I misunderstood! 😄

Wilfredwongkc committed:
Add GUI feature but only replies input

Wilfredwongkc committed:
Add additional work on GUI

Wilfredwongkc committed:
Merge branch 'branch-Level-10'

Wilfredwongkc committed:
Remove unwanted code

Wilfredwongkc committed:
Add assertions and fix bug in printing for the find command

Find command: make find command display correct

The find command did not work as it was printing, it will not be
shown on the GUI as the input required for it is a string.

Let's,
* make the find command case insensitive
* abstract the commands into classes

Wilfredwongkc committed:
Refactor code by replacing magic literals

Wilfredwongkc committed:
Merge pull request #1 from Wilfredwongkc/branch-A-CodeQuality

Refactor code by replacing magic literals

Wilfredwongkc committed:
Fix bugs

Wilfredwongkc committed:
Add back exception handling

Wilfredwongkc committed:
Add help function and change images

Wilfredwongkc committed:
Merge branch 'branch-A-Assertions'

Wilfredwongkc committed:
Refactor code to remove arrow head shape. Place code in Parser class

Wilfredwongkc committed:
Remove unused code and add javadocs for some methods

Wilfredwongkc committed:
Remove more unused comment and some weird spaces

Wilfredwongkc committed:
Remove unwanted spaces at the end and add javadocs for parse method

Wilfredwongkc committed:
Remove unused code

Wilfredwongkc committed:
Remove help comments from the guide

Wilfredwongkc committed:
Add new exception handling for no task number behind command

Wilfredwongkc committed:
Add representative screenshot

Wilfredwongkc committed:
Add README.md

Wilfredwongkc committed:
Remove lines and fixed table not showing bug

Wilfredwongkc committed:
Add launcher class and change gradle settings to create jar file

Wilfredwongkc committed:
Change javadocs comments by adding 's'

Wilfredwongkc committed:
Add space to make table appear

Wilfredwongkc committed:
Remove plugin in build gradle file to for jar to work

Wilfredwongkc committed:
Fix bug of jar not being able to run
